### ISSUE TITLE
LSP Extensions support

### ISSUE THREAD

**Issue by fannheyward** [fannheyward/coc-rust-analyzer#256]
https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/lsp-extensions.md

- [ ] SnippetTextEdit, https://github.com/rust-analyzer/rust-analyzer/pull/4494, #252
- [ ] joinLines, https://github.com/rust-analyzer/rust-analyzer/pull/4557
- [ ] SSR, https://github.com/rust-analyzer/rust-analyzer/pull/4560

**Comment by oblitum** [comment]
fwiw, at #252, the two reports of testers at the end pointed that snippet placeholder selection isn't working.

**Comment by fannheyward** [comment]
@oblitum selection in same file works as expect, but fails on another file, this also happens with RA+VSCode extension, https://github.com/rust-analyzer/rust-analyzer/issues/4551.

I've disabled selection for different file.

**Comment by oblitum** [comment]
I only tested it in same file, didn't work for me. Meaning, just writing `foo()` and expanding it I get to the end of the file as [reported before](https://github.com/fannheyward/coc-rust-analyzer/pull/252#issuecomment-631838323), which I though was the [same report by @TimoFreiberg](https://github.com/fannheyward/coc-rust-analyzer/pull/252#issuecomment-633011552), but I'm not sure he was reporting the trivial test now.

**Comment by fannheyward** [comment]
@oblitum can't reproduce the error in same file tests, selection will not go to the end of file.

**Comment by oblitum** [comment]
k, fwiw, this is my setup:

```
vim version: NVIM v0.4.3
node version: v14.3.0
coc.nvim version: 0.0.78-3c774fbfa5
coc-rust-analyzer version: 0.7.2
rust-analyzer version: fbb8b88
coc.nivm directory: /home/francisco/.vim/plugged/coc.nvim
platform: linux
```

If @TimoFreiberg could chime in to state whether [this report](https://github.com/fannheyward/coc-rust-analyzer/pull/252#issuecomment-633011552) was about a trivial test in the same file or not, it would be great.

**Comment by fannheyward** [comment]
@oblitum reproduced your issue, very very weird behavior:

```rust
fn main() {
    foo();
} // -- no new line EOF
```

this will add `foo` function with selection on `|}|`.

```rust
fn main() {
    foo();
}
// -- with new line EOF
```

this will add `foo` function with selection on `todo!()`.

**Comment by oblitum** [comment]
@fannheyward I confirm it's just like that.

**Comment by TimoFreiberg** [comment]
Sorry for the delay, yeah I was also talking about that.
Interesting that it seems to work as expected when there's a newline at the end - rustfmt removed that on save so I didn't notice that case.

**Comment by linde12** [comment]
@fannheyward realize that i'm late here but i just wanted to chime in on the decorations/inlay hints - currently it is supported for chained calls, but it'd be super nice to have for inferred types as well so that you could do e.g. `let (tx, rx) = std::sync::mpsc::channel();` and analyzer would add decorations like `let (tx: Sender<T>, rx: Receiver<T>) = std::sync::mpsc::channel();`

**Comment by fannheyward** [comment]
@linde12 we can't do inlay hints in vim https://github.com/fannheyward/coc-rust-analyzer/issues/224

**Comment by linde12** [comment]
@fannheyward yeah i realize this but in nvim we could with the decoration API right? (i assume thats how inlayHints work now for chained calls?)

**Comment by pierreN** [comment]
Is https://rust-analyzer.github.io/manual.html#workspace-symbol anywhere planed or did I miss something by any chance? thanks

**Comment by fannheyward** [comment]
@pierreN try `:CocList symbols`

**Comment by fannheyward** [comment]
LSP 3.16 will add code action resolving support, need to switch to upstream https://github.com/rust-analyzer/rust-analyzer/pull/6521

**Comment by GopherJ** [comment]
LSP 3.16 has file operations, which can be super useful for rust projects: https://github.com/rust-analyzer/rust-analyzer/pull/7009

**Comment by resolritter** [comment]
@fannheyward "Range hover" is marked as done but the commit was reverted (https://github.com/fannheyward/coc-rust-analyzer/commit/d6b4f338a09d7c29d64815bab80a64d35f55fdaf), could you please contextualize? Happy to help with the implementation if you need it

**Comment by fannheyward** [comment]
@resolritter yes, I reverted and pushed it to another branch https://github.com/fannheyward/coc-rust-analyzer/tree/feat/range-hover, the problem is I did find a better way to make Range hover work with nnoremap K. Do you have any ideas?

**Comment by resolritter** [comment]
> @resolritter yes, I reverted and pushed it to another branch https://github.com/fannheyward/coc-rust-analyzer/tree/feat/range-hover, the problem is I did find a better way to make Range hover work with nnoremap K. Do you have any ideas?

You're using [getSelectedRange](https://github.com/fannheyward/coc-rust-analyzer/commit/f6092a7a30b9ecd352ebef93479c07c6064cbc79#) which [expects Vim to be in visual mode](https://github.com/neoclide/coc.nvim/blob/54e9b983d3c871c46d6bfaae4f7df126ae123fee/src/__tests__/handler/selectionRange.test.ts#L61), however `:<C-U>` kind of mappings (e.g. `xmap <Plug>(coc-codeaction-selected)`) exit visual mode before executing the range and therefore that check will not pass.

While you could use `<cmd>` mappings to avoid exiting visual mode, exiting is needed to get the visual mode range with `'<` and `'>` like [here](https://github.com/neoclide/coc.nvim/blob/c805bf0a63c67936cf9475cb47cd078280959071/src/workspace.ts#L648) (as far as I'm aware it's not possible to do that while staying in visual mode, you have to exit it first for those marks' positions to be updated).

So, in short, the solution is to make the mapping work like `xmap <Plug>(coc-codeaction-selected)`. [Here's the doCodeAction code](https://github.com/neoclide/coc.nvim/blob/c805bf0a63c67936cf9475cb47cd078280959071/src/handler/codeActions.ts#L85) for reference and you'll see that it fetches the visual range using the marks' positions as mentioned.

**Comment by fannheyward** [comment]
@resolritter sorry for the delay and thank you for your help, I'm busy today, will check this next week.

**Comment by fannheyward** [comment]
For range hover, you need mappings like `vnoremap <silent> K <cmd>call CocActionAsync('doHover')<CR>` to work.

**Comment by AceofSpades5757** [comment]
Any plans to implement _Inlay Hints_ now that Vim supports virtual text?

**Comment by fannheyward** [comment]
@AceofSpades5757 this will add in coc.nvim
